### PR TITLE
feat: upgrade to react 17

### PR DESCRIPTION
**Pull request by blaky**
### Proposed behaviour

Upgrade to React 17.

### Current behaviour

It is currently on React 16

### Checklist

- [X] Commits follow our style guide
- [X] Related issues linked in commit messages if required
- [ ] Screenshots are included in the PR if useful
- [ ] All themes are supported if required
- [X] Unit tests added or updated if required
- [ ] Cypress automation tests added or updated if required
- [ ] Storybook added or updated if required
- [ ] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [X] Typescript `d.ts` file added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Additional context

I upgraded the library to React 17, there is only one unit test that I couldn't fix:
`Toast › modal manager › when component unmounts › it is removed from modal manager`
I don't fully understand how the toasts work with the modal manager, so I would like to kindly ask you to take a look at that test. All other tests pass and the components seem to work correctly. 

### Testing instructions

Regression test

### CODE REVIEW

**Comment by blaky** [comment]
React 17 handles the error logging somewhat differently, so most of the changes I did was related to the `console.error` expectations. I had to add a new `expect` function so the code can be tested the same way as before, with the existing expectation conditions.

**Comment by blaky** [comment]
It fixes https://github.com/Sage/carbon/issues/4708

**Comment by mkrds** [comment]
@blaky Hey Bence, thanks for your PR :)

I've added two commits, first one fixes the issue that was caught by the toast unit test you mentioned, it's caused by this change in React 17 https://pl.reactjs.org/blog/2020/08/10/react-v17-rc.html#effect-cleanup-timing

I've also added second commit, take a loot at form-field.spec.js and see an alternative way of handling the changes in React 17 error logging. Using that approach there's no need to extend the expect method, especially considering that since we're refactoring the codebase to TypeScript we'll be able to get rid of propTypes. That will also render your new `toHaveBeenCalledWithFormattedErrorMessage` not needed anymore after our refactor is done. 

What do you think about it?

**Comment by blaky** [comment]
Hey @mkrds 

Thank you for the changes, they look good to me. I am going to squash all changes and force push to the branch.

File: package.json [hunk 1]
```diff
@@ -152,9 +152,9 @@
     "mockdate": "^2.0.2",
     "prettier": "^2.1.2",
     "raf": "^3.4.0",
-    "react": "^16.14.0",
-    "react-dom": "^16.14.0",
-    "react-test-renderer": "^16.14.0",
+    "react": "^17.0.0",
```
Comment (mkrds) [inline package.json:155]:
```suggestion
    "react": "^17.0.2",
```

File: package.json [hunk 2]
```diff
@@ -152,9 +152,9 @@
...
+    "react-dom": "^17.0.0",
```
Comment (mkrds) [inline package.json:156]:
```suggestion
    "react-dom": "^17.0.2",
```

File: package.json [hunk 3]
```diff
@@ -152,9 +152,9 @@
...
+    "react-test-renderer": "^17.0.0",
```
Comment (mkrds) [inline package.json:157]:
```suggestion
    "react-test-renderer": "^17.0.2",
```

File: package.json [hunk 4]
```diff
@@ -90,20 +91,20 @@
...
     "@testing-library/react": "^12.1.3",
     "@types/enzyme": "^3.10.3",
-    "@types/enzyme-adapter-react-16": "^1.0.5",
     "@types/invariant": "^2.2.35",
     "@types/jest": "^26.0.19",
     "@types/lodash": "^4.14.179",
     "@types/node": "10.12.18",
-    "@types/react": "16.9.56",
-    "@types/react-dom": "16.9.8",
+    "@types/react": "^17.0.0",
```
Comment (mkrds) [inline package.json:98]:
```suggestion
    "@types/react": "^17.0.2",
```

File: package.json [hunk 5]
```diff
@@ -90,20 +91,20 @@
...
+    "@types/react-dom": "^17.0.0",
```
Comment (mkrds) [inline package.json:99]:
```suggestion
    "@types/react-dom": "^17.0.2",
```

**Comment by blaky** [comment]
Hi @mkrds , thanks again for the work in this PR!

@DipperTheDan , The PR is now ready, I squashed all changes into a single commit.

**Comment by nicktitchmarsh** [comment]
We have to put a hold on the merge of this PR due to some issues with MFEs. We need to wait for those projects to complete initial work in order to take this upgrade. Otherwise we will block them from taking other bug fixes and changes after this commit.

**Review by edleeks87 — COMMENTED** [review]
@blaky thanks for raising this, I'm happy to approve the proposed changes when the conflicts been resolved

**Comment by jamime** [comment]
ℹ️ We're still assessing upgrade paths in various internal projects. We should know more by the end of this week.

**Review by mkrds — DISMISSED** [review]

**Review by edleeks87 — DISMISSED** [review]

**Review by mkrds — APPROVED** [review]

**Review by edleeks87 — APPROVED** [review]

**Comment by jamime** [comment]
ℹ️ We will be issuing a breaking change, carbon will no longer support React 17.

**Pull request merged**